src/awkward/_nplikes/typetracer.py

```diff
@@ -294,60 +209,34 @@ def _length_after_slice(slice, original_length):
 
 
 class TypeTracerArray(NDArrayOperatorsMixin, ArrayLike):
-    @classmethod
-    def from_array(cls, array, dtype: np.dtype | None = None) -> Self:
-        """
-        Args:
-            array: array-like object, e.g. np.ndarray, #ak.index.Index
-            dtype: dtype of returned #ak._typetracer.TypeTracerArray
-
-        Returns an #ak._typetracer.TypeTracerArray that describes the type information
-        of the given array.
-        """
-        if isinstance(array, index.Index):
-            array = array.data
-
-        # not array-like, try and cast to a NumPy array
-        elif not hasattr(array, "shape"):
-            array = numpy.array(array)
-
-        if array.dtype == np.dtype("O"):
-            raise ak._errors.wrap_error(
-                ValueError(
-                    f"bug in Awkward Array: attempt to construct `TypeTracerArray` "
-                    f"from a sequence of non-primitive types: {array}"
-                )
+    def __new__(cls, *args, **kwargs):
+        raise wrap_error(
+            TypeError(
+                "internal_error: the `TypeTracer` nplike's `TypeTracerArray` object should never be directly instantiated"
```

Comment:
This protects from people accidentally importing `TypeTracerArray` and operating upon it naively.